invisible_cities/io/dst_io.py

```diff
@@ -29,53 +29,52 @@ def load_dsts(dst_list, group, node):
+    arr_types = arr.dtype
+    if set(arr_types.names) != set(table_types.names):
+        raise TableMismatch(f'dataframe differs from already existing table structure')
+    for name in arr_types.names:
+        if arr_types[name].name == 'object':
+            max_str_length = max(map(len, arr[name]))
+            if max_str_length > table_types[name].itemsize:
+                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
+        elif arr_types[name] != table_types[name]:
+            warnings.warn(f'dataframe numeric types not consistent with the table existing ones', UserWarning)
+
+
 def store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name : str, table_name : str, compression : str='ZLIB4', descriptive_string : [str]="", str_col_length : int=32) -> None:
-    if len(df) == 0:
-        warnings.warn(f'dataframe is empty', UserWarning)
     if group_name not in h5out.root:
         group = h5out.create_group(h5out.root, group_name)
-    arr = df.to_records(index=False)
     group = getattr(h5out.root, group_name)
 
+    arr = df.to_records(index=False)
+
     if table_name not in group:
-        tabledef = _make_tabledef(df.dtypes, str_col_length=str_col_length)
+        tabledef = _make_tabledef(arr.dtype, str_col_length=str_col_length)
         table    =  make_table(h5out,
                                group       = group_name,
                                name        = table_name,
                                fformat     = tabledef,
                                description = descriptive_string,
                                compression = compression)
-        #hack to transform object numpy dtype to string
-        def _cast_type(dtype : np.dtype):
-            if dtype != 'O': return dtype
-            else:            return f'S{str_col_length}'
-        data_types = np.dtype([(col, _cast_type(arr[col].dtype)) for col in arr.dtype.names])
     else:
         table = getattr(group, table_name)
-        data_types = table.dtype
 
-    if not np.array_equal(df.columns, table.colnames):
-        raise TableMismatch(f'dataframe differs from already existing table structure')
-
-    #check which columns are of type object and throw warning if length exceeds
-    for colname in df.columns:
-        if (df[colname].dtype.name == 'object'):
-            if df[colname].str.len().max() > data_types[colname].itemsize:
-                warnings.warn(f'dataframe contains strings longer than allowed', UserWarning)
-        else:
-            if arr.dtype[colname] != data_types[colname]:
-                    warnings.warn(f'dataframe numeric types not consistent with the table existing ones', UserWarning)
-
-    arr = arr.astype(data_types)
-    table.append(arr)
-    table.flush()
+    data_types = table.dtype
+    if len(arr) == 0:
+        warnings.warn(f'dataframe is empty', UserWarning)
+    else:
+        _can_cast(arr, data_types)
```

Comment:
I suggest a better name like `_check_castability` or something like that. The use of "can" suggests that it returns a boolean value.